design/resource-dependency.md

```diff
@@ -0,0 +1,348 @@
+      - componentName: my-rds
+        dataOutputs:
+        - name: mysql-conn-secret
+          fieldPath: status.connSecret
+    ```
+
+2. Add a field `valueFrom.dataOutputName` in Component ParameterValue to specify the dependency on data outputs:
+
+    ```yaml
+    parameterValues:
+    - name: <parameter-name>
+      valueFrom:
+        dataOutputName: <dataoutput-object-name>
+    ```
+
+    For example, web app's dependency on RDS's database secret could be specified as:
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: database-secret # workload's fieldPaths needs to be filled
+        fieldPaths:
+        - spec.connSecret
+    ---
+
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        parameterValues:
+        - name: database-secret
+          valueFrom:
+            # Specify the dependency on a data output.
+            # The component's workload won't be created until the data output has returned value to fill workload's fieldPaths. 
+            dataOutputName: mysql-conn-secret
+
+      - componentName: my-rds
+        dataOutputs:
+        - name: mysql-conn-secret
+          fieldPath: status.connSecret
+    ```
+
+    In this case, the OAM runtime will automatically build a dependency from `my-web` to `my-rds`.
+
+    It will wait until `my-rds` object's field `status.connSecret` has value:
+
+    ```yaml
+    apiVersion: example.io/v1alpha1
+    kind: RDSInstance
+    metadata:
+      name: my-rds
+    status:
+      connSecret: mysql-conn # OAM runtime will wait for this field to have value
+    ```
+
+    Once the dependency has been satisfied, the OAM runtime will create `my-web` workload with
+    the corresponding field paths patched with values from data outputs.
+    
+    ```yaml
+    kind: MyWorkload
+    metadata:
+      name: my-web
+    spec:
+      connSecret: mysql-conn # patched by OAM runtime
+    ```
+
+3. Similary, add fields `dataOutputs` and `parameters` in Trait configuration to specify the dependency on data output objects.
+
+    ```yaml
+    parameters:
+    - valueFrom:
+        dataOutputName: <dataoutput-object-name>
+      toFieldPaths: <field-paths-to-fill>
+    dataOutputs:
+    - name: <dataoutput-name>
+      fieldPath: <object-field-path>
+    ```
+
+    For example, we have an etcd backup trait that relies on the data output from an etcd cluster resource.
+    We can specify it as:
+
+    ```yaml
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-etcd-cluster
+        dataoutputs:
+        - name: etcd-conn-secret
+          fieldpath: "status.connSecret"
+
+        traits:
+        - trait:
+            apiVersion: example.io/v1alpha1
+            kind: EtcdBackup
+            metadata:
+              name: my-etcd-backup
+          parameters:
```

Comment:
why is this a `parameter` instead of `parameterValue`? Also, I don't seem to see a new `parameter` field in the `traitDefinition` object

Reply:
I think this should be called `parameterValues` so that the keyword `parameter` only means "declare a parameter" in OAM. IMO, it is too confusing for `parameter` to mean `declare` (in Component) or `assign` (in Trait) depends on which resources it appears in the OAM spec.

Reply:
Going to change this to data input fields